src/mlpack/methods/decision_tree/random_categorical_split_impl.hpp

```diff
@@ -0,0 +1,256 @@
+/**
+ * @file methods/decision_tree/random_categorical_split_impl.hpp
+ * @author Adarsh Santoria
+ *
+ * Implementation of the RandomCategoricalSplit categorical split class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_DECISION_TREE_RANDOM_CATEGORICAL_SPLIT_IMPL_HPP
+#define MLPACK_METHODS_DECISION_TREE_RANDOM_CATEGORICAL_SPLIT_IMPL_HPP
+
+namespace mlpack {
+
+// Overload used in classification.
+template<typename FitnessFunction>
+template<bool UseWeights, typename VecType, typename LabelsType,
+         typename WeightVecType>
+double RandomCategoricalSplit<FitnessFunction>::SplitIfBetter(
+    const double bestGain,
+    const VecType& data,
+    const size_t numCategories,
+    const LabelsType& labels,
+    const size_t numClasses,
+    const WeightVecType& weights,
+    const size_t minimumLeafSize,
+    const double minimumGainSplit,
+    arma::vec& splitInfo,
+    AuxiliarySplitInfo& /* aux */)
+{
+  const double epsilon = 1e-7; // Tolerance for floating-point errors.
+
+  // Count the number of elements in each potential child.
+  arma::Col<size_t> counts(numCategories, arma::fill::zeros);
+  for (size_t i = 0; i < data.n_elem; ++i)
+    counts[(size_t) data[i]]++;
+  
+  // Calculate the gain of the split.  First we have to calculate the labels
+  // that would be assigned to each child.
+  arma::uvec childPositions(numCategories, arma::fill::zeros);
+  std::vector<arma::Row<size_t>> childLabels(numCategories);
+  std::vector<arma::Row<double>> childWeights(numCategories);
```

Comment:
I have not gone deeply through the code but almost certainly at a glance it is not necessary to do this.  Ideally you should be calculating the loss without copying the data---and these objects amount to fairly expensive copies of the data.